sys/linux/socket.txt

```diff
@@ -161,6 +177,20 @@ cmsghdr {
 	data		array[int8]
 } [align_ptr]
 
+type cmsghdr_unaligned[SOL, TYP, DATA] {
+	cmsg_len	len[parent, intptr]
+	cmsg_level	const[SOL, int32]
+	cmsg_type	const[TYP, int32]
+	data		DATA
+} [packed]
+
+# For __sock_cmsg_send
+cmsghdr_sock [
```

Comment:
This does not look right. We do need `[align_ptr]` somewhere, because messages are supposed to be aligned on `intptr`. Check CMSG_SPACE definition. For example, mark will occupy 20 bytes, and the next message won't be aligned properly.

Reply:
Do you mean that a [align_ptr] cmsghdr should be add in socket.txt? Just like what you mentioned on the mailing? But it's not used by any syscall and get the "unused type" fail. Or should i add another sendmsg$* for it?

Reply:
Here is what I sketched on the mailing list:
```
type cmsghdr_t[SOL, TYP, DATA] {
  unaligned cmsghdr_unaligned[SOL, TYP, DATA]
} [align_ptr]

type cmsghdr_unaligned[SOL, TYP, DATA] {
  len len[parent, intptr]
  level const[SOL, int32]
  type const[TYP, int32]
  data DATA
} [packed]

Then e.g. unix socket cmsgs can be described as:
cmsghdr_un [
  rights cmsghdr_t[SOL_SOCKET, SCM_RIGHTS, array[fd]]
  cred cmsghdr_t[SOL_SOCKET, SCM_CREDENTIALS, cmsg_un_cred]
] [varlen]

cmsg_un_cred {
  pid pid
  uid uid
  gid gid
}
```

We need something like this. cmsghdr_t is used by cmsghdr_un union.